File: library/src/main/java/com/mobilesolutionworks/android/app/WorksActivity.java
package com.mobilesolutionworks.android.app;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;

import com.mobilesolutionworks.android.app.controller.WorksControllerManager;

/**
 * Activity host for WorksController.
 * <p>
 * Created by yunarta on 19/11/15.
 */
public class WorksActivity extends AppCompatActivity {

//    private SparseArray<FragmentTrackInfo> mTrackInfoMap;

    /**
     * Controller manager.
     */
    private WorksControllerManager mController;

    public WorksActivity() {
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        WorksControllerManager.Loader loader = (WorksControllerManager.Loader) getSupportLoaderManager().initLoader(0, null, new WorksControllerManager.LoaderCallbacks(this));
        mController = loader.getController();
    }

    @Override
    protected void onPostCreate(@Nullable Bundle savedInstanceState) {
        super.onPostCreate(savedInstanceState);
        mController.getLifecycleHook().onRestoreInstanceState(savedInstanceState);
    }

    /**
     * Get controller manager to create individual controller.
     *
     * @return controller manager.
     */
    public WorksControllerManager getControllerManager() {
        return mController;
    }

    @Override
    public void onResume() {
        super.onResume();
        mController.getLifecycleHook().dispatchResume();
    }

    @Override
    public void onPause() {
        super.onPause();
        mController.getLifecycleHook().dispatchPause();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        mController.getLifecycleHook().dispatchSaveInstanceState(outState);
        super.onSaveInstanceState(outState);
    }
}

File: library/src/main/kotlin/com/mobilesolutionworks/android/app/WorksActivity.kt
package com.mobilesolutionworks.android.app

import android.content.res.Configuration
import android.os.Bundle
import android.support.v7.app.AppCompatActivity

import com.mobilesolutionworks.android.app.controller.WorksControllerManager

/**
 * Activity host for WorksController.
 *
 * Created by yunarta on 19/11/15.
 */
open class WorksActivity : AppCompatActivity(), WorkControllerHost {

    var loader: WorksControllerManager.ControllerManager? = null

    /**
     * Get controller manager to create individual controller.
     * @return controller manager.
     */
    override val controllerManager: WorksControllerManager
        get() {
            return loader!!.controller
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        loader = supportLoaderManager.initLoader(0, null, WorksControllerManager.ControllerManagerLoaderCallbacks(this)) as WorksControllerManager.ControllerManager
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        controllerManager.getLifecycleHook().onRestoreInstanceState(savedInstanceState)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        controllerManager.getLifecycleHook().onConfigurationChanged(newConfig)
    }

    public override fun onResume() {
        super.onResume()
        controllerManager.getLifecycleHook().dispatchResume()
    }

    public override fun onPause() {
        super.onPause()
        controllerManager.getLifecycleHook().dispatchPause()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        controllerManager.getLifecycleHook().dispatchSaveInstanceState(outState)
        super.onSaveInstanceState(outState)
    }
}
